### DISCUSSION TITLE
Extend the app to allow interactive conversations AKA Chatbot

### DISCUSSION THREAD

**Discussion by ahassoun** [shridarpatil/frappe_whatsapp#44] [Ideas]
Hello!

What would be _your_ way to extend the app to cover the following features?

1. Prompt users to enter their details, each message will fill a field in a Doctype, where their phone number is the unique primary key. They will be prompted to fill in some info (Enter your first name... Enter your last name... etc).
2. Answer questions related to them, where their phone number is, again, the searchable key as well as the authenticator. Messages like: What are my tasks for today? Will check the sender's phone number and select records accordingly.
3. Based on a certain keyword/option (e.g., Talk to an Advisor), the next messages will be forwarded to an external web service, and also send back the response from the service to WhatsApp. Something like an external Chatbot that has a webhook capability, and instead of this Chatbot taking over the WhatsApp number, it will receive prompts from Frappe, this app. This can remain open for a time window.

The question might be related to the Frappe framework, and I am looking for your opinion here. What does it take to (listen) to the `message` field for all new WhatsApp Message documents, and based on some keywords, trigger some server script? Or are there other, more straightforward approaches?

**Comment by shridarpatil** [comment]
I dint get your 1st point. But 2nd and 3rd you can achieve it by using hooks on message doctype where you can add your custom logic.

**Reply by ahassoun** [reply]
1st point: Imagine a DocType named (Guest) that collects guests' information. The primary key here will be a phone number. When new Guest chats with us, they are prompted to fill in their details, one field at a message.

So I'll have one controller that listens to .. `after_insert`? And checks for strings... based on the string, it may redirect to another function, right?

Is `after_insert` the right hook?
